---
title: "Gene_Geneset"
author: "Fan"
date: "2020年4月20日"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
setwd("D:\\data_work\\R_sources\\my_code\\Gene_Geneset")
```

## 生信
### 差异分析之limma
差异分析当然推荐limma流程，标准流程的输入数据为RNA-seq的counts矩阵或芯片数据。
该方法也支持从FPKM/TPM或GSVA数据开始差异分析（直接将数据矩阵扔进lmFit即可，可以参考GSVA文档示例browseVignettes("GSVA")中的第9页）。
输入counts数据可以对数据进行voom处理（虽然不知道这种处理的优势在哪）。值得注意的是，常用的另外两个工具包edgeR和DESeq2都不支持FPKM等标准化后的数据作为输入，仅limma支持。

RNA-Seq的的limma分析标准流程示例：[R pipeline to analyze RNA-seq data](http://bioinf.wehi.edu.au/RNAseqCaseStudy/)和相应的[code](http://bioinf.wehi.edu.au/RNAseqCaseStudy/code.R)
```{r limma}
my_limma<-function(data,group_info,adjust.method="fdr",data_type="counts",
                   p.value.cutoff=1,fold.change.cutoff=1){
  require(dplyr)
  require(limma)
  require(edgeR)
  
  case_info_id<-tibble(case_id=colnames(data)[-1])
  case_info<-inner_join(case_info_id,group_info,by=c("case_id"=colnames(group_info)[1]))
  if (all(as.vector(unlist(group_info[,1]))!=colnames(data)[-1])){
    cat("Re-arranged!\n")
  }
  
  group_origin<-case_info[colnames(group_info)[2]] %>% unlist() %>% as.vector() %>% as.factor()
  group<-as.factor(paste("G",as.character(as.numeric(group_origin)),sep=""))
  design <- model.matrix(~ 0+group)
  colnames(design) <- levels(group)
  
  expr<-data[-1]
  gene_names<-data[,1] %>% unlist() %>% as.vector()
  if (length(gene_names)!=length(unique(gene_names))) { cat("Duplicate genes!\n") }
  expr_mat<-as.matrix(expr)
  row.names(expr_mat)<-gene_names
  
  # filter low expression
  data_med<-apply(expr_mat, 1, median)
  low_index<-which(data_med<0.001)
  expr_mat<-expr_mat[-low_index,]
  if (length(low_index)>0){
    cat("fitered low expression!\n")
  }
  
  # log2 transform
  ex <- expr_mat
  qx <- as.numeric(quantile(ex, c(0., 0.25, 0.5, 0.75, 0.99, 1.0), na.rm=T))
  LogC <- (qx[5] > 100) ||
    (qx[6]-qx[1] > 50 && qx[2] > 0) ||
    (qx[2] > 0 && qx[2] < 1 && qx[4] > 1 && qx[4] < 2)
  if (LogC) {
    ex[which(ex <= 0)] <- NaN
    expr_mat <- log2(ex)
    cat("log transforming...\n")
  }
  
  if (data_type=="counts"){
    cat("using method for counts...\n")
    # fit the model
    x <- DGEList(counts=expr_mat, genes=row.names(expr_mat))
    # perform voom normalization
    y <- voom(x,design,plot=TRUE)
    # cluster libraries
    # ? plotMDS(y,xlim=c(-2.5,2.5))
    # fit linear model and assess differential expression
    fit <- lmFit(y, design)
    
  }
  if (data_type %in% c("fpkm","tpm","rpkm")){
    cat("using method for FPKM/RPKM/TPM...\n")
    # fit the model
    fit <- lmFit(expr_mat, design)
  }
  if ((data_type %in% c("counts","fpkm","tpm","rpkm"))==FALSE){
    cat("data type not supported!\n")
    return(0)
  }
  cont.matrix <- makeContrasts(G2-G1, levels=design)
  fit2 <- contrasts.fit(fit, cont.matrix)
  fit2 <- eBayes(fit2)
  limma_results <- topTable(fit2,adjust.method=adjust.method,coef=1,p.value=p.value.cutoff,
                              lfc=log(fold.change.cutoff,2),number=50000,sort.by = 'logFC')
  return(as_tibble(limma_results,rownames="gene"))
}
```
注意事项  
**是否需要比较矩阵**
可参考这个讨论：[差异分析是否需要比较矩阵](https://github.com/bioconductor-china/basic/blob/master/makeContrasts.md)
```{r limma_2,message=FALSE,warning=FALSE}
# method 1
library(CLL)
data(sCLLex)
library(limma)
design=model.matrix(~factor(sCLLex$Disease))
fit=lmFit(sCLLex,design)
fit=eBayes(fit)
options(digits = 4)
topTable(fit,coef=2,adjust='BH')

# method 2
design=model.matrix(~0+factor(sCLLex$Disease))
colnames(design)=c('progres','stable')
fit=lmFit(sCLLex,design)
cont.matrix=makeContrasts('progres-stable',levels = design)
fit2=contrasts.fit(fit,cont.matrix)
fit2=eBayes(fit2)
options(digits = 4)
topTable(fit2,adjust='BH')
```

### 差异分析之edgeR
官方文档中明确指出，edgeR仅支持RNA-Seq原始的counts数据，不支持标准化处理后的数据（如FPKM/RPKM/TPM等），略微兼容RSEM数据，但是不推荐。具体介绍请参考：edgeRUsersGuide()。如下为封装的代码函数（暂时未进行测试，慎用）。
```{r edgeR}
my_edgeR<-function(){
  require(dplyr)
  require(edgeR)
  
  case_info_id<-tibble(case_id=colnames(data)[-1])
  case_info<-inner_join(case_info_id,group_info,by=c("case_id"=colnames(group_info)[1]))
  if (paste(as.vector(unlist(group_info[,1])),collapse = "")!=
      paste(colnames(data)[-1],collapse = "")){
    cat("grouping...\n")
  }
  
  group_origin<-case_info[colnames(group_info)[2]] %>% unlist() %>% as.vector() %>% as.factor()
  group<-as.factor(paste("G",as.character(as.numeric(group_origin)),sep=""))
  
  expr<-data[-1]
  gene_names<-data[,1] %>% unlist() %>% as.vector()
  if (length(gene_names)!=length(unique(gene_names))) { cat("Duplicate genes!\n") }
  expr_mat<-as.matrix(expr)
  row.names(expr_mat)<-gene_names
  
  # main code
  y <- DGEList(counts=expr_mat,group=group)
  y <- calcNormFactors(y) # default method is TMM
  y <- estimateCommonDisp(y)
  y <- estimateTagwiseDisp(y)
  et <- exactTest(y,pair = levels(group))
  ordered_tags <- topTags(et, n=100000)
  temp <- ordered_tags$table
  edgeR_results = temp[is.na(temp$FDR)==FALSE,]
}
```

### 差异表达分析之DESeq2
官方文档中明确指出，DESeq2的表达数据需为原始counts数据，不支持FPKM等标准化后的数据。HT-seq的count是可以的，请参考文档：Rstudio的help处搜索DESeq2。如下为代码示例
```{r DESeq2, message=FALSE,warning=FALSE}
library("pasilla")
pasCts <- system.file("extdata",
                      "pasilla_gene_counts.tsv",
                      package="pasilla", mustWork=TRUE)
pasAnno <- system.file("extdata",
                       "pasilla_sample_annotation.csv",
                       package="pasilla", mustWork=TRUE)
cts <- as.matrix(read.csv(pasCts,sep="\t",row.names="gene_id"))
coldata <- read.csv(pasAnno, row.names=1)
coldata <- coldata[,c("condition","type")]
rownames(coldata) <- sub("fb", "", rownames(coldata))
# all(rownames(coldata) %in% colnames(cts))
## [1] TRUE
# all(rownames(coldata) == colnames(cts))
## [1] FALSE
cts <- cts[, rownames(coldata)]
# all(rownames(coldata) == colnames(cts))
## [1] TRUE
library("DESeq2")
dds <- DESeqDataSetFromMatrix(countData = cts,
                              colData = coldata,
                              design = ~ condition)
dds <- DESeq(dds)
res <- results(dds)
resOrdered <- res[order(res$pvalue),]
sum(res$padj < 0.1, na.rm=TRUE)
res.df<-as.data.frame(res)
head(res.df)
```

### 火山图
**简单火山图**
```{r simpleVocano}
my_simple_volcanoPlot<-function(logFC, pvalue, logFC.cutoff=1, pvalue.cutoff=0.05,
                                xlab="Log2(Fold change)", ylab="-Log10(Pvalue)",
                                main="",xlim=NULL,ylim=NULL){
  
  if (is.null(xlim)) xlim<-max(abs(logFC))
  if (is.null(ylim)) ylim<-max(-log10(pvalue))
  data<-tibble(logFC=logFC, pvalue=pvalue,
               DEG=rep('notDEG',length(logFC)),
               colors=rep('grey',length(logFC)))
  # set colors
  data$DEG[data$logFC > logFC.cutoff & data$pvalue < pvalue.cutoff] = 'Up'
  data$DEG[data$logFC < -logFC.cutoff & data$pvalue < pvalue.cutoff] = 'Down'
  data$colors = factor(data$DEG,levels=c("Down","notDEG","Up"),
                       labels=c("darkgreen","grey","red"))
  
  # draw scatter plot 
  plot(data$logFC, -log10(data$pvalue), pch=16,
       col=as.character(data$colors),
       xlab=xlab, ylab=ylab, cex=0.5, main=main, xlim=c(-xlim,xlim),ylim=c(0,ylim))
  abline(h=-log(pvalue.cutoff,10),col="black",lwd=1)
  abline(v=c(logFC.cutoff,-logFC.cutoff),col="black",lwd=1)
}
```

## 富集分析

富集分析包括ORA和GSEA两种，当然GSEA比较可靠。
富集分析的可视化，其实不管富集分析的结果来自ORA还是GSEA分析，某些可视化其实是可以通用的。

主要使用的工具包：clusterProfiler、enrichplot、topGO、GOplot、pathview。
clusterProfiler是个很不错的包，可参考：browseVignettes("clusterProfiler")

```{r enrichment, message=FALSE,warning=FALSE, cache=TRUE}
library(clusterProfiler)
library(org.Hs.eg.db)
data(geneList, package="DOSE")
gene <- names(geneList)[abs(geneList) > 2]
ego <- enrichGO(gene          = gene,
                universe      = names(geneList),
                OrgDb         = org.Hs.eg.db,
                ont           = "CC",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.01,
                qvalueCutoff  = 0.05,
                readable      = TRUE)

kk <- enrichKEGG(gene         = gene,
                 organism     = 'hsa',
                 pvalueCutoff = 0.05)

kk2 <- gseKEGG(geneList     = geneList,
               organism     = 'hsa',
               nPerm        = 1000,
               minGSSize    = 120,
               pvalueCutoff = 0.05,
               verbose      = FALSE)
```




```{r enrichment_plot}
barplot(ego, showCategory=8)
dotplot(ego)
emapplot(ego)

## categorySize can be scaled by 'pvalue' or 'geneNum'
cnetplot(ego, categorySize="pvalue", foldChange=geneList)
cnetplot(ego, foldChange=NULL, circular = TRUE, colorEdge = TRUE)
goplot(ego)
heatplot(ego, foldChange=NULL)

# GSEA 也可以画这个图，展示基因集之间的关系
png("GOgraph.png",height = 8000,width=5000,res=500,unit="px")
plotGOgraph(ego)
dev.off()

# 这个图要再本地查看高清版才行
library(EBImage)
pic <- readImage("GOgraph.png")
display(pic)

gseaplot(kk2, geneSetID = "hsa04145")

# using the code nextline can bowse the KEGG graph
# browseKEGG(kk, 'hsa04110')

library("pathview")
# library(pathview) 去掉引号似乎会出错
hsa04110 <- pathview(gene.data  = geneList,
                     pathway.id = "hsa04110",
                     species    = "hsa",
                     limit      = list(gene=max(abs(geneList)), cpd=1))
# head(hsa04110)

library(EBImage)
pic = readImage("hsa04110.pathview.png")
display(pic)
```

# GOgraph, out.width, out.height

```{r}
image("GOgraph.png")
```

</br>
&nbsp;

[GOplot](https://wencke.github.io/)官方教程讲得很清楚。

注意：如下和弦图中的基因和基因集是进一步选定给出的，而不是差异分析和富集分析的直接结果。
这个图其实也可以用于GSEA的结果展示！
```{r GOplot, message=FALSE,warning=FALSE}
library(GOplot)
david<- EC$david
genelist<- EC$genelist
circ <- circle_dat(EC$david, EC$genelist)
genes<- EC$genes
process<-EC$process

chord <- chord_dat(data=circ, genes=EC$genes,  process=EC$process)

GOChord(chord, space = 0.02, limit = c(3, 5),
        gene.order = 'logFC', gene.space = 0.25, gene.size = 2,
        lfc.col=c('firebrick3', 'white','royalblue3'),
        ribbon.col=brewer.pal(length(EC$process), "Set3"))+
  theme(plot.margin=unit(c(0,5,0,5),'lines'))+
  theme(legend.title = element_text(size=6))+
  theme(legend.text = element_text(size=6))
  
GOCluster(circ, EC$process, clust.by = 'logFC', term.width = 2)+
  theme(legend.title = element_text(size=6))+
  theme(legend.text = element_text(size=6))

GOCluster(circ, EC$process, clust.by = 'term', lfc.col = c('darkgoldenrod1', 'black', 'cyan1'))+
  theme(legend.title = element_text(size=6))+
  theme(legend.text = element_text(size=6))

```
[WGCNA](https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/)
代码比较复杂，写在了另一个代码库里。




##其他生信工具
estimate包是否只适用于芯片数据？

[immunedeconv](https://icbi-lab.github.io/immunedeconv/articles/immunedeconv.html)、[相应的github](https://github.com/icbi-lab/immunedeconv)、[及其示例](https://icbi-lab.github.io/immunedeconv/articles/detailed_example.html)：各类工具的集合

[MCP-counter](https://cit.ligue-cancer.net/tools/)

[EPIC](https://gfellerlab.shinyapps.io/EPIC_1-1/)

[EPIC_github](https://github.com/GfellerLab/EPIC)




